src/Orleans/Threading/CancellationTokenManager.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal static class CancellationTokenManager
+    {
+       /// <summary>
+       /// wraps found cancellation tokens into instances of type CancellationTokenWrapper
+       /// </summary>
+       /// <param name="arguments"></param>
+        public static void WrapCancellationTokens(object[] arguments)
+        {
+            if (arguments == null) return;
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                var argument = arguments[i];
+                if (argument is CancellationToken)
+                {
+                    arguments[i] = WrapCancellationToken((CancellationToken)argument);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Wraps cancellation token into CancellationTokenWrapper
+        /// </summary>
+        /// <param name="ct"> Cancellation token to be wrapped</param>
+        /// <returns>CancellationTokenWrapper</returns>
+        public static CancellationTokenWrapper WrapCancellationToken(CancellationToken ct)
+        {
+            var tokenId = Guid.NewGuid();
+            return new CancellationTokenWrapper(tokenId, ct);
+        }
+
+        /// <summary>
+        /// Retrieves cancellation token from holder grain
+        /// </summary>
+        /// <param name="tokenId"> Id of token to be retrieved</param>
+        /// <returns></returns>
+        public static Task<CancellationToken> GetCancellationToken(Guid tokenId)
+        {
+            return RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenId).GetCancellationToken();
+        }
+
+        internal static void RegisterTokenCallbacks(CancellationToken ct, Guid tokenId)
+        {
+            if (!ct.CanBeCanceled) return;
+            ct.Register((f) =>
+            {
+                Cancel(tokenId).ContinueWith(task =>
+                {
+                    if (task.IsFaulted)
+                    {
+                        var logger = TraceLogger.GetLogger("CancellationTokenManager", TraceLogger.LoggerType.Runtime);
+                        if (logger.IsWarning)
+                        {
+                            logger.Warn(ErrorCode.Runtime_Error_100332, "Remote token cancellation failed", task.Exception);
+                        }
+                    }
+                });
```

Comment:
Would it make sense to use `TaskContinuationOptions.OnlyOnFaulted` here instead of explicit `task.Faulted`? I'm not entirely sure if it would improve inlining etc.